### PR TITLE
EN-6888: Implement GeoRegion validation

### PR DESCRIPTION
**Pull request by aerust**
Adds validation for `georegion_match_on_point` and
`georegion_match_on_string` computation strategies.

Co-written by: roark.brewster@socrata.com

### CODE REVIEW

File: src/main/scala/com/socrata/computation_strategies/GeoRegionMatchStrategy.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package com.socrata.computation_strategies
+
+import com.rojoma.json.v3.codec.{Path, DecodeError, JsonDecode}
+import com.socrata.soql.types.{SoQLPoint, SoQLType}
+
+object GeoRegionMatchStrategy {
+
+
+  def validate[ColumnName : JsonDecode, Schema: JsonDecode](strategyType: StrategyType,
```
Comment (meganschneider) [inline src/main/scala/com/socrata/computation_strategies/GeoRegionMatchStrategy.scala]: nit: spacing around colons is inconsistent

File: src/main/scala/com/socrata/computation_strategies/GeoRegionMatchStrategy.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+          }
+
+          // validate the parameters
+          optParameters match {
+            case Some(obj) => JsonDecode.fromJValue[Schema](obj) match {
+              case Right(_) => None // success
+              case Left(error) => error match {
+                case DecodeError.MissingField(field, Path.empty) => Some(MissingParameter(field))
+                case _ => Some(InvalidStrategyParameters(error))
+              }
```
Comment (rjmac) [inline src/main/scala/com/socrata/computation_strategies/GeoRegionMatchStrategy.scala]: I'd probably flatten this match into
```scala
JsonDecode.blahblah match {
  case Right(_) => None // Success
  case Left(DecodeError.MisisngField(field, Path.empty)) => Some(MisisngParameter(field))
  case Left(otherError) => Some(InvalidStrategyParemeters(otherError))
}
```

File: src/main/scala/com/socrata/computation_strategies/GeoRegionMatchOnStringComputationStrategy.scala [hunk 3]
```diff
@@ -1,8 +1,36 @@
...
+import com.rojoma.json.v3.util.{AutomaticJsonDecodeBuilder, AutomaticJsonEncodeBuilder}
+import com.socrata.soql.types.{SoQLText, SoQLNumber, SoQLType}
 
+case class GeoRegionMatchOnStringParameterSchema[T,U](region: T, column: U, primaryKey: Option[U])
+
+object GeoRegionMatchOnStringParameterSchema  extends ParameterSchema {
+  implicit def encoder[T : JsonEncode, U : JsonEncode] = AutomaticJsonEncodeBuilder[GeoRegionMatchOnStringParameterSchema[T,U]]
+  implicit def decoder[T : JsonDecode, U : JsonDecode] = AutomaticJsonDecodeBuilder[GeoRegionMatchOnStringParameterSchema[T,U]]
+
+  override def requiredFields = Seq("region", "column")
```
Comment (meganschneider) [inline src/main/scala/com/socrata/computation_strategies/GeoRegionMatchOnStringComputationStrategy.scala]: do we also want a list for optional fields, and to enforce that nothing outside of those required+optional params exists? may be automagical with json; haven't looked through All The Things yet

**Comment by rjmac** [comment]
Seems ok to me; in general I'd prefer to have flatter matches (that is, instead of
```scala
  val Thing(x,y,z) = structuredThing
  x match {
    case X => y match {
       case Y1 => ...
       case Y2 => ...
    }
    case ...
  }
```
I like
```scala
  structuredThing match {
     case Thing(X, Y1, _) => ...
     case Thing(X, Y2, _) => ...
     ...
  }
```
if possible.  It isn't always, but despite the slight repetition it makes for easier-to-read code.

**Comment by meganschneider** [comment]
Tests, both positive and negative, for the formatting/validation? Are they in cheetah-land instead?

**Comment by roarkb** [comment]
Writing Cheetah tests is the very next thing on our todo list :)

File: src/main/scala/com/socrata/computation_strategies/GeoRegionMatchOnStringComputationStrategy.scala (hunk 3 above)
Comment (aerust) [inline src/main/scala/com/socrata/computation_strategies/GeoRegionMatchOnStringComputationStrategy.scala]: I don't think we are worried too much about extra fields in json. But that does seem to be cores approach:

```
{
  "code" : "invalid_request",
  "error" : true,
  "message" : "Invalid Request: Deserialization error at line 13 column 13: No property 'extra' for object type 'ViewColumn'"
}
```

**Comment by meganschneider** [comment]
lgtm, pending robert thumbsup :)

**Comment by aerust** [comment]
(Robert said it looks good)

**Pull request merged**